### PR TITLE
better at-least-once semantics regarding commiting position

### PR DESCRIPTION
**Pull request by osheroff**
previously there was some possibility of losing data if one partition was master-less, while others were fine.  Now we keep a list of outstanding messages, and set our binlog position as the last continuous
position that's completed.  

@vanchi-zendesk @dadah89 , do you remember talking about binlog positions?  The old code always did `setSync` -- ie after processing DDL the replicator would force the binlog position to be written out to the database -- but I'm really unclear why I thought that was necessary.  Maybe it was that previously it was very expensive (or damaging?) to replay a schema update?  

@zendesk/rules /cc @dasch (some of this is prep-work for independent consumer positions)

### CODE REVIEW

**Comment by dasch** [comment]
Should the inflight message buffer be bounded?

Otherwise looks good :+1:

File: src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+			this.isComplete = false;
+		}
+	}
+
+	private LinkedList<InflightMessage> list;
+	private HashMap<String, InflightMessage> hash;
+
+	public InflightMessageList() {
+		this.list = new LinkedList<>();
+		this.hash = new HashMap<>();
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java]: LinkedHashMap?

Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java]: Yeah, ok.  The whole thing with modifying iterators and such is pretty confusing but it looks like that'd work.

File: src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		BinlogPosition completeUntil = null;
+		Iterator<InflightMessage> iterator = this.linkedMap.values().iterator();
+
+		while ( iterator.hasNext() ) {
+			InflightMessage msg = iterator.next();
+			if ( !msg.isComplete )
+				break;
+
+			completeUntil = msg.position;
+			iterator.remove();
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java]: this works, although IMO it's kinda oblique from the code; `values().iterator()` actually iterates in the map itself, so the `remove()` call removes the node safely.

> https://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html#values()

**Comment by osheroff** [comment]
@dasch, there's an implicit bound on the message buffer, which is however big the outstanding kafka buffer is.  Probably best to just leave the bounding there rather than trying to impose another limit.

File: src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java (hunk 2 above)
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java]: ok, looks like the standard way: http://stackoverflow.com/questions/223918/iterating-through-a-collection-avoiding-concurrentmodificationexception-when-re

But I am sure you are missing some foldLeft flatMap action here? :-D

**Comment by vanchi-zendesk** [comment]
👍

**Comment by dasch** [comment]
> @dasch, there's an implicit bound on the message buffer, which is however big the outstanding kafka buffer is.  Probably best to just leave the bounding there rather than trying to impose another limit.

:+1:

**Comment by dadah89** [comment]
@osheroff  sorry I'm late to comment, but just a question: what happens if a message fails to send? does it keep retrying until sending it? or do we risk getting the binlog position stuck?

I remember talking about/reviewing smt related to replicator/binlog position, maybe the comments in this commit will help? https://github.com/zendesk/maxwell/commit/569ce244 (smt about replicator thread being restarted and picking up the right position?)

**Comment by osheroff** [comment]
@dadah89 the binlog position can and should get stuck if we don't deliver the message.. but I should look into what actually happens.  It's not all that great to keep writing way ahead for one partition and then have to backtrack and send all the data over again.

**Comment by dadah89** [comment]
worst case, you could get stuck until there is no more binlog file =) would be nice to have some limit to how much it can lag behind

**Comment by osheroff** [comment]
well like I said to @dasch, eventually the kafka buffer for that partition is gonna fill up and the replicator will stop.  Hard to know the "right" amount to backpressure is really, but I should try to get a jepsen-y thing setup locally and see how maxwell reacts in various situations.

**Comment by dadah89** [comment]
ah didn't realize it was a similar question. If it stops once the buffer fills up, seems fine to me

**Pull request merged**